docker/docker_client.go

```diff
@@ -133,16 +133,34 @@ func dockerCertDir(ctx *types.SystemContext, hostPort string) string {
 		return ctx.DockerCertPath
 	}
 	var hostCertDir string
-	if ctx != nil && ctx.DockerPerHostCertDirPath != "" {
-		hostCertDir = ctx.DockerPerHostCertDirPath
-	} else if ctx != nil && ctx.RootForImplicitAbsolutePaths != "" {
-		hostCertDir = filepath.Join(ctx.RootForImplicitAbsolutePaths, systemPerHostCertDirPath)
-	} else {
-		hostCertDir = systemPerHostCertDirPath
+	for _, systemPerHostCertDirPath := range systemPerHostCertDirPaths {
+		if ctx != nil && ctx.DockerPerHostCertDirPath != "" {
+			hostCertDir = ctx.DockerPerHostCertDirPath
+		} else if ctx != nil && ctx.RootForImplicitAbsolutePaths != "" {
+			hostCertDir = filepath.Join(ctx.RootForImplicitAbsolutePaths, systemPerHostCertDirPath)
+		} else {
+			hostCertDir = systemPerHostCertDirPath
+		}
+
+		if pathExists(filepath.Join(hostCertDir, hostPort)) {
+			break
+		}
 	}
 	return filepath.Join(hostCertDir, hostPort)
 }
 
+// pathExists returns true if path exists otherwise false
+func pathExists(path string) bool {
+	_, err := os.Stat(path)
+	if err == nil {
+		return true
+	}
+	if os.IsNotExist(err) {
+		return false
+	}
+	return true
```

Comment:
This is a bit suspect; the error may be ”a parent directory (`/etc/containers/certs.d`) is unreadable”, for example (note that `/etc/docker/certs.d` is only readable by `root` by default).

Treating other errors as `true` may still be the right thing to do, but it’s not immediately obvious to me that it is.

Reply:
Should the error be returned instead? This way the user might have an idea of why its failing.

Reply:
Constraints:

- If the directory exists _and is readable_, it must be used (the happy “things work” path)
- If the directory definitely does not exist, we must try the other one
- If a permission is denied accessing the `certs.d` directory or the `hostPort` subdirectory, we must not fail entirely (the “unprivileged user with root-only certificates” case); we should try with the default TLS configuration.

Lower-priority concerns:
- I/O errors should not be silently ignored; preferably we should abort and return the I/O error.
- If a permission is denied accessing `/` or `/etc`, or other unlikely cases, it would be nice to treat them same as I/O errors — but we can’t easily distinguish this from the “`certs.d` unreadable” case, which is more important.
- _At least_ the debug logs should make it clear what is going on and why (see how `SetupCertificates` logs its decisions)

In theory it could be possible for `/etc/containers/certs.d` to exist and be unaccessible, and `/etc/docker/certs.d` to exist and be usable; I don’t think this inconsistent case is worth worrying about too much. Same for `/etc/containers/certs.d/$hostPort` existing but being unaccessible (i.e. `Stat` returning no error but `syscall.Access` reporting the directory as being not accessible).

---

With the limited error information easily available in the `os` package, the above probably translates to

- `err == nil`: use the path
- `os.IsNotExist(err)`: try the next path
- `os.IsPermission(err)`: log this, try the next path
- other `err != nil` cases: fail, return the `err`

---

Given the above semantics requirements, and especially the need to log the reasons for the decisions, it seems to me that the decision can’t be neatly abstracted/packaged into something `certs.d`-agnostic like `pathExists`; it may be cleaner to inline the decision into the caller, even if that makes it a big bigger.

Reply:
fixed.